### PR TITLE
Vultr integration

### LINKED ISSUE #10705: Vultr Service Discovery Support
## Proposal
**Use case. Why is this important?**

At https://www.vultr.com/ we have had a large number of users ask for us to integrate with Prometheus so they could use our platform for their service discovery. We may also end up using this internally for various projects.

I already have a branch with our changes implemented https://github.com/ddymko/prometheus/commit/1003fba104ba830f15aeb2aeecad6f1806787591 and I would be open to maintaining our provider integration.

I figured it would be better to create a feature request first instead of sending a PR without discussion first. Let me know if there is anything that you may need from me!

### PR DESCRIPTION
**Pull request by ddymko**
Vultr integration which resolves #10705

Signed-off-by: David Dymko dymkod@gmail.com

### CODE REVIEW

**Comment by ddymko** [comment]
I'll address the failures in another commit

**Comment by ddymko** [comment]
I keep running into this failure when I run tests locally. I am not quite sure how to resolve it didn't want to just change the value without checking if this is a misconfiguration on my end.


```
=== RUN   TestLoadConfig
    config_test.go:1098: 
        	Error Trace:	config_test.go:1098
        	Error:      	Not equal: 
        	            	expected: &config.Config{GlobalConfig:config.GlobalConfig{ScrapeInterval:15000000000, ScrapeTimeout:10000000000, EvaluationInterval:30000000000, QueryLogFile:"", ExternalLabels:labels.Labels{labels.Label{Name:"foo", Value:"bar"}, labels.Label{Name:"monitor", Value:"codelab"}}}, AlertingConfig:config.AlertingConfig{AlertRelabelConfigs:[]*relabel.Config(nil), AlertmanagerConfigs:config.AlertmanagerConfigs{(*config.AlertmanagerConfig)(0x4849260)}}, RuleFiles:[]string{"testdata/first.rules", "testdata/my/*.rules"}, ScrapeConfigs:[]*config.ScrapeConfig{(*config.ScrapeConfig)(0x484bec0), (*config.ScrapeConfig)(0x484c020), (*config.ScrapeConfig)(0x484c180), (*config.ScrapeConfig)(0x484c440), (*config.ScrapeConfig)(0x484c5a0), (*config.ScrapeConfig)(0x484c700), (*config.ScrapeConfig)(0x484c860), (*config.ScrapeConfig)(0x484c9c0), (*config.ScrapeConfig)(0x484cb20), (*config.ScrapeConfig)(0x484cc80), (*config.ScrapeConfig)(0x484cde0), (*config.ScrapeConfig)(0x484cf40), (*config.ScrapeConfig)(0x484a8c0), (*config.ScrapeConfig)(0x484aa20), (*config.ScrapeConfig)(0x484ab80), (*config.ScrapeConfig)(0x484ace0), (*config.ScrapeConfig)(0x484ae40), (*config.ScrapeConfig)(0x484afa0), (*config.ScrapeConfig)(0x484b100), (*config.ScrapeConfig)(0x484b260), (*config.ScrapeConfig)(0x484b3c0), (*config.ScrapeConfig)(0x484b520), (*config.ScrapeConfig)(0x484b680), (*config.ScrapeConfig)(0x484b7e0), (*config.ScrapeConfig)(0x484b940), (*config.ScrapeConfig)(0x484baa0), (*config.ScrapeConfig)(0x484bc00), (*config.ScrapeConfig)(0x484bd60)}, StorageConfig:config.StorageConfig{ExemplarsConfig:(*config.ExemplarsConfig)(nil)}, TracingConfig:config.TracingConfig{ClientType:"grpc", Endpoint:"localhost:4317", SamplingFraction:0, Insecure:false, TLSConfig:config.TLSConfig{CAFile:"", CertFile:"testdata/valid_cert_file", KeyFile:"testdata/valid_key_file", ServerName:"", InsecureSkipVerify:true, MinVersion:0x0}, Headers:map[string]string{"foo":"bar"}, Compression:"gzip", Timeout:5000000000}, RemoteWriteConfigs:[]*config.RemoteWriteConfig{(*config.RemoteWriteConfig)(0x4849ea0), (*config.RemoteWriteConfig)(0x4849ae0)}, RemoteReadConfigs:[]*config.RemoteReadConfig{(*config.RemoteReadConfig)(0x4848420), (*config.RemoteReadConfig)(0x48486c0)}}
        	            	actual  : &config.Config{GlobalConfig:config.GlobalConfig{ScrapeInterval:15000000000, ScrapeTimeout:10000000000, EvaluationInterval:30000000000, QueryLogFile:"", ExternalLabels:labels.Labels{labels.Label{Name:"foo", Value:"bar"}, labels.Label{Name:"monitor", Value:"codelab"}}}, AlertingConfig:config.AlertingConfig{AlertRelabelConfigs:[]*relabel.Config(nil), AlertmanagerConfigs:config.AlertmanagerConfigs{(*config.AlertmanagerConfig)(0xc000b1f100)}}, RuleFiles:[]string{"testdata/first.rules", "testdata/my/*.rules"}, ScrapeConfigs:[]*config.ScrapeConfig{(*config.ScrapeConfig)(0xc000199b80), (*config.ScrapeConfig)(0xc000199e40), (*config.ScrapeConfig)(0xc000172160), (*config.ScrapeConfig)(0xc000172840), (*config.ScrapeConfig)(0xc0001729a0), (*config.ScrapeConfig)(0xc000172b00), (*config.ScrapeConfig)(0xc000172c60), (*config.ScrapeConfig)(0xc000172dc0), (*config.ScrapeConfig)(0xc000172f20), (*config.ScrapeConfig)(0xc000173080), (*config.ScrapeConfig)(0xc0001731e0), (*config.ScrapeConfig)(0xc000173340), (*config.ScrapeConfig)(0xc0001734a0), (*config.ScrapeConfig)(0xc000173600), (*config.ScrapeConfig)(0xc000173760), (*config.ScrapeConfig)(0xc0001738c0), (*config.ScrapeConfig)(0xc000173a20), (*config.ScrapeConfig)(0xc000173b80), (*config.ScrapeConfig)(0xc000173ce0), (*config.ScrapeConfig)(0xc000173e40), (*config.ScrapeConfig)(0xc0004a62c0), (*config.ScrapeConfig)(0xc0004a6420), (*config.ScrapeConfig)(0xc0004a7600), (*config.ScrapeConfig)(0xc0002c89a0), (*config.ScrapeConfig)(0xc0000fc000), (*config.ScrapeConfig)(0xc0000fc160), (*config.ScrapeConfig)(0xc0000fc2c0), (*config.ScrapeConfig)(0xc0000fc420)}, StorageConfig:config.StorageConfig{ExemplarsConfig:(*config.ExemplarsConfig)(nil)}, TracingConfig:config.TracingConfig{ClientType:"grpc", Endpoint:"localhost:4317", SamplingFraction:0, Insecure:false, TLSConfig:config.TLSConfig{CAFile:"", CertFile:"testdata/valid_cert_file", KeyFile:"testdata/valid_key_file", ServerName:"", InsecureSkipVerify:true, MinVersion:0x0}, Headers:map[string]string{"foo":"bar"}, Compression:"gzip", Timeout:5000000000}, RemoteWriteConfigs:[]*config.RemoteWriteConfig{(*config.RemoteWriteConfig)(0xc000165b80), (*config.RemoteWriteConfig)(0xc000165e00)}, RemoteReadConfigs:[]*config.RemoteReadConfig{(*config.RemoteReadConfig)(0xc0001f7380), (*config.RemoteReadConfig)(0xc0001f75f0)}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -3970,3 +3970,3 @@
        	            	       FollowRedirects: (bool) true,
        	            	-      EnableHTTP2: (bool) false
        	            	+      EnableHTTP2: (bool) true
        	            	      },
        	Test:       	TestLoadConfig
--- FAIL: TestLoadConfig (0.02s)


Expected :&config.Config{GlobalConfig:config.GlobalConfig{ScrapeInterval:15000000000, ScrapeTimeout:10000000000, EvaluationInterval:30000000000, QueryLogFile:"", ExternalLabels:labels.Labels{labels.Label{Name:"foo", Value:"bar"}, labels.Label{Name:"monitor", V ...

Actual   :&config.Config{GlobalConfig:config.GlobalConfig{ScrapeInterval:15000000000, ScrapeTimeout:10000000000, EvaluationInterval:30000000000, QueryLogFile:"", ExternalLabels:labels.Labels{labels.Label{Name:"foo", Value:"bar"}, labels.Label{Name:"monitor", V ...

<Click to see difference>
```

File: discovery/vultr/vultr.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+	vultrInstanceLabelVCPU         = vultrInstanceLabel + "vcpu_count"
+	vultrInstanceLabelMemory       = vultrInstanceLabel + "ram_bytes"
+	vultrInstanceLabelBandwidth    = vultrInstanceLabel + "bandwidth_bytes"
+	vultrInstanceLabelDisk         = vultrInstanceLabel + "disk_gb"
+	separator                      = ","
+
+	// vultrInstanceLabelSize = vultrInstanceLabel + "size"
+)
+
+// DefaultSDConfig is the default Linode SD configuration.
```
Comment (roidelapluie) [inline discovery/vultr/vultr.go:60]:
```suggestion
// DefaultSDConfig is the default Vultr SD configuration.
```

File: discovery/vultr/vultr.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	vultrInstanceLabelTags         = vultrInstanceLabel + "tags"
+	vultrInstanceLabelHostname     = vultrInstanceLabel + "hostname"
+	vultrInstanceLabelServerStatus = vultrInstanceLabel + "server_status"
+	vultrInstanceLabelVCPU         = vultrInstanceLabel + "vcpu_count"
+	vultrInstanceLabelMemory       = vultrInstanceLabel + "ram_bytes"
+	vultrInstanceLabelBandwidth    = vultrInstanceLabel + "bandwidth_bytes"
+	vultrInstanceLabelDisk         = vultrInstanceLabel + "disk_gb"
+	separator                      = ","
+
+	// vultrInstanceLabelSize = vultrInstanceLabel + "size"
```
Comment (roidelapluie) [inline discovery/vultr/vultr.go:57]: Do we need this?

File: .gitignore [hunk 3]
```diff
@@ -28,3 +28,5 @@ npm_licenses.tar.bz2
 /.build
 
 /**/node_modules
+
+.idea/
```
Comment (roidelapluie) [inline .gitignore:32]: We do not add editor's gitignore in Prometheus. Please consider using a global .gitignore on your computer.

File: config/config_test.go [hunk 4]
```diff
@@ -996,6 +997,31 @@ var expectedConf = &Config{
...
+			HTTPClientConfig: config.DefaultHTTPClientConfig,
+
+			ServiceDiscoveryConfigs: discovery.Configs{
+				&vultr.SDConfig{
+					HTTPClientConfig: config.HTTPClientConfig{
+						Authorization: &config.Authorization{
+							Type:        "Bearer",
+							Credentials: "abcdef",
+						},
+						FollowRedirects: true,
```
Comment (roidelapluie) [inline config/config_test.go:1018]:
```suggestion
						FollowRedirects: true,
						EnableHTTP2: true,
```

**Comment by roidelapluie** [comment]
There is a few commented code lines that needs removal; and this PR requires a rebase. I will move further with our account.

File: config/config_test.go (hunk 4 above)
Comment (ddymko) [reply config/config_test.go:1042]: ah thanks! I wasn't sure if this needed to be added or not

**Comment by roidelapluie** [comment]
I can not test yet as we are not yet 'integrators' but I think this SD does not work if you check 'No public IPv4' when creating an instance.

**Comment by ddymko** [comment]
@roidelapluie Your vultr account should be good to go.


Hmm I didn't consider users not setting an IPv4 address

**Comment by roidelapluie** [comment]
Can you also update plugins.yml and plugins/plugins.go

**Comment by ddymko** [comment]
Umm.....I'll follow up on that - I'll also get the plugins update

**Comment by ddymko** [comment]
@roidelapluie This is expected behavior it deploys a `100.` address for https://datatracker.ietf.org/doc/html/rfc6598

**Comment by roidelapluie** [comment]
Then I propose we renaime public_ipv4 and public_ipv6 labels: main_ip and main_ipv6 to match what the API provides

File: discovery/vultr/vultr.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+	vultrInstanceLabel             = model.MetaLabelPrefix + "vultr_instance_"
+	vultrInstanceLabelID           = vultrInstanceLabel + "id"
+	vultrInstanceLabelLabel        = vultrInstanceLabel + "label"
+	vultrInstanceLabelOS           = vultrInstanceLabel + "os"
+	vultrInstanceLabelOSID         = vultrInstanceLabel + "os_id"
+	vultrInstanceLabelRegion       = vultrInstanceLabel + "region"
+	vultrInstanceLabelPlan         = vultrInstanceLabel + "plan"
+	vultrInstanceLabelMainIP       = vultrInstanceLabel + "main_ip"
+	vultrInstanceLabelMainIPV6     = vultrInstanceLabel + "main_ipv6"
+	vultrInstanceLabelPrivateIPv4  = vultrInstanceLabel + "internal_ip"
```
Comment (roidelapluie) [inline discovery/vultr/vultr.go:46]:
```suggestion
	vultrInstanceLabelInternalIP  = vultrInstanceLabel + "internal_ip"
```

**Review by roidelapluie — COMMENTED** [review]
Thanks! Can you please add the documentation in docs/configuration/configuration.md ?

File: discovery/vultr/vultr.go (hunk 5 above)
Comment (ddymko) [reply discovery/vultr/vultr.go:46]: Completely forgot about docs! Added those in and also updated the label for bandwidth to match the API response

File: docs/configuration/configuration.md [hunk 6]
```diff
@@ -2538,6 +2538,83 @@ tls_config:
...
+* `__meta_vultr_instance_features` : List of features that are available to the instance.
+* `__meta_vultr_instance_tags` : List of tags associated with the instance.
+* `__meta_vultr_instance_hostname` : The hostname for this instance.
+* `__meta_vultr_instance_server_status` : The server health status.
+* `__meta_vultr_instance_vcpu_count` : Number of vCPUs.
+* `__meta_vultr_instance_ram_bytes` : The amount of RAM in MB.
+* `__meta_vultr_instance_disk_gb` : The size of the disk in GB.
+* `__meta_vultr_instance_allowed_bandwidth` : Monthly bandwidth quota in GB.
+
+```yaml
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:2569]: Indentation is not correct in this block.

File: discovery/vultr/vultr.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+	vultrInstanceLabelMainIP       = vultrInstanceLabel + "main_ip"
+	vultrInstanceLabelMainIPV6     = vultrInstanceLabel + "main_ipv6"
+	vultrInstanceLabelInternalIP   = vultrInstanceLabel + "internal_ip"
+	vultrInstanceLabelFeatures     = vultrInstanceLabel + "features"
+	vultrInstanceLabelTags         = vultrInstanceLabel + "tags"
+	vultrInstanceLabelHostname     = vultrInstanceLabel + "hostname"
+	vultrInstanceLabelServerStatus = vultrInstanceLabel + "server_status"
+	vultrInstanceLabelVCPU         = vultrInstanceLabel + "vcpu_count"
+	vultrInstanceLabelMemory       = vultrInstanceLabel + "ram_bytes"
+	vultrInstanceLabelBandwidth    = vultrInstanceLabel + "allowed_bandwidth"
```
Comment (roidelapluie) [inline discovery/vultr/vultr.go:53]:
```suggestion
	vultrInstanceLabelBandwidth    = vultrInstanceLabel + "allowed_bandwidth_gb"
```

**Comment by ghubz** [comment]
I updated to v2.36 of prometheus, set up the VULTR config but when I Try to reload the config I get: 

line 34: field credentials not found in type vultr.plain

**Comment by ghubz** [comment]
That's how I had it and it didn't work. I had to add: 

  vultr_sd_configs:
    - bearer_token:

Btw please consider adding **power_status**, currently you cannot know if a machine is powered off or not. server_status is completely bogus as it returns OK even if a server is stopped.

**Comment by roidelapluie** [comment]
Based on you error message you were missing indentation

**Pull request merged**